File: generated_code/python/solution_420_2.py
from typing import List

class Solution:
    def strongPasswordChecker(self, password: str) -> int:
        n = len(password)
        
        # 1. Count missing character types (lowercase, uppercase, digit)
        has_lower = any(c.islower() for c in password)
        has_upper = any(c.isupper() for c in password)
        has_digit = any(c.isdigit() for c in password)
        missing_types = (1 if not has_lower else 0) + (1 if not has_upper else 0) + (1 if not has_digit else 0)
        
        # 2. Find all repeating sequences of length >= 3
        # We store the lengths of these repeating groups
        repeating_groups = []
        i = 0
        while i < n:
            j = i
            while j < n and password[j] == password[i]:
                j += 1
            length = j - i
            if length >= 3:
                repeating_groups.append(length)
            i = j
            
        # --- Case 1: Short password (len < 6) ---
        # We only need to insert characters. Each insertion can also break a repeating triplet.
        # Total steps is the max of required insertions to reach length 6 and missing types.
        if n < 6:
            return max(6 - n, missing_types)
            
        # --- Case 2: Standard length (6 <= len <= 20) ---
        # We can fix triplets purely by replacements. 
        # Each triplet of length L needs L // 3 replacements.
        elif n <= 20:
            replacements = sum(length // 3 for length in repeating_groups)
            return max(replacements, missing_types)
            
        # --- Case 3: Overly long password (len > 20) ---
        # We MUST delete exactly (n - 20) characters.
        # Strategy: Use deletions effectively to reduce the total number of required replacements.
        # A repeating group of length L requires L // 3 replacements.
        # - If L % 3 == 0, 1 deletion reduces replacements by 1 (e.g., length 3 -> 2, replacements 1 -> 0).
        # - If L % 3 == 1, 2 deletions reduce replacements by 1 (e.g., length 4 -> 2, replacements 1 -> 0).
        # - If L % 3 == 2, 3 deletions reduce replacements by 1 (e.g., length 5 -> 2, replacements 1 -> 0).
        else:
            deletions_needed = n - 20
            
            # Priority 1: Groups with length % 3 == 0 (costs 1 deletion to save 1 replacement)
            for i in range(len(repeating_groups)):
                if deletions_needed <= 0:
                    break
                if repeating_groups[i] % 3 == 0:
                    repeating_groups[i] -= 1
                    deletions_needed -= 1
                    
            # Priority 2: Groups with length % 3 == 1 (costs 2 deletions to save 1 replacement)
            for i in range(len(repeating_groups)):
                if deletions_needed <= 0:
                    break
                if repeating_groups[i] % 3 == 1:
                    # We can use up to 2 deletions for this group
                    rem = min(deletions_needed, 2)
                    repeating_groups[i] -= rem
                    deletions_needed -= rem
                    
            # Priority 3: Groups with length % 3 == 2 (costs 3 deletions to save 1 replacement)
            for i in range(len(repeating_groups)):
                if deletions_needed <= 0:
                    break
                if repeating_groups[i] >= 3:
                    # Each 3 deletions reduces length and saves 1 replacement
                    # We look at how many replacements are left that can be removed by 3 deletions
                    possible_3s_to_remove = repeating_groups[i] // 3
                    rem = min(deletions_needed, possible_3s_to_remove * 3)
                    repeating_groups[i] -= rem
                    deletions_needed -= rem
            
            # Calculate final replacements needed after optimization
            replacements = sum(length // 3 for length in repeating_groups if length >= 3)
            
            # Total steps = deletions made + max of remaining replacements or missing types
            return (n - 20) + max(replacements, missing_types)

File: generated_code/cpp/solution_420_2.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int strongPasswordChecker(string password) {
        int n = password.length();
        
        // Check for missing types: lowercase, uppercase, and digit
        bool has_lower = false, has_upper = false, has_digit = false;
        for (char c : password) {
            if (islower(c)) has_lower = true;
            if (isupper(c)) has_upper = true;
            if (isdigit(c)) has_digit = true;
        }
        int missing_types = (has_lower ? 0 : 1) + (has_upper ? 0 : 1) + (has_digit ? 0 : 1);
        
        // Count lengths of repeating character sequences
        vector<int> repeats;
        for (int i = 0; i < n; ) {
            int j = i;
            while (j < n && password[j] == password[i]) {
                j++;
            }
            int len = j - i;
            if (len >= 3) {
                repeats.push_back(len);
            }
            i = j;
        }
        
        // Case 1: Password is too short
        // We only need insertions. Insertions can also fix repetitions and missing types.
        if (n < 6) {
            return max(6 - n, missing_types);
        }
        
        // Case 2: Password has valid length [6, 20]
        // We only need replacements. Replacements can resolve both repetitions and missing types.
        if (n <= 20) {
            int replace_needed = 0;
            for (int len : repeats) {
                replace_needed += len / 3;
            }
            return max(replace_needed, missing_types);
        }
        
        // Case 3: Password is too long
        // Deletions are mandatory to reach length 20. 
        // We use deletions strategically to reduce the number of replacements needed.
        int delete_needed = n - 20;
        int overlength = delete_needed;
        
        // Priority 1: Reduce repeats where len % 3 == 0. (1 deletion saves 1 replacement)
        for (int& len : repeats) {
            if (overlength > 0 && len % 3 == 0) {
                len -= 1;
                overlength -= 1;
            }
        }
        
        // Priority 2: Reduce repeats where len % 3 == 1. (2 deletions save 1 replacement)
        for (int& len : repeats) {
            if (overlength > 0 && len % 3 == 1) {
                int rem = min(overlength, 2);
                len -= rem;
                overlength -= rem;
            }
        }
        
        // Priority 3: Reduce remaining repeats (3 deletions save 1 replacement)
        for (int& len : repeats) {
            if (overlength > 0 && len >= 3) {
                int max_del = len - 2; // Keep at least 2 characters to avoid creating new patterns
                int rem = min(overlength, max_del);
                len -= rem;
                overlength -= rem;
            }
        }
        
        // Calculate remaining replacements needed after optimal deletions
        int replace_needed = 0;
        for (int len : repeats) {
            if (len >= 3) {
                replace_needed += len / 3;
            }
        }
        
        // Total steps = required deletions + max of required replacements or missing types
        return delete_needed + max(replace_needed, missing_types);
    }
};